### PR TITLE
Feature/77 catalog

### LINKED ISSUE #77: Federated catalog
MVD participants must deploy the EDC federated catalog, and populate it with the catalog from other participants.

- Deploy EDC federated catalog `catalog-cache` within EDC
- When deploying each participant, write out its URL to a shared file storage account
- In EDC, build a custom extension that periodically reloads endpoints from the shared file storage account, and updates catalog cache nodes

### LINKED ISSUE #85: Federated catalog
MVD participants must deploy the EDC federated catalog, and populate it with the catalog from other participants.

- Deploy EDC federated catalog `catalog-cache` within EDC
- When deploying each participant, write out its URL to a shared file storage account
- In EDC, build a custom extension that periodically reloads endpoints from the shared file storage account, and updates catalog cache nodes

### PR DESCRIPTION
**Pull request by algattik**
Added federated catalog service, and a custom extension to populate it from JSON files. Adapted Terraform deployment to produce these files, and restart instances after deployment. See included ADR for details.

Added a system test to verify that the federated catalog is indeed populated.

Closes agera-edc/MinimumViableDataspaceFork#85

### CODE REVIEW

File: deployment/terraform/main.tf [hunk 1]
```diff
@@ -194,3 +220,18 @@ resource "azurerm_storage_blob" "did" {
...
+resource "local_file" "catalog_entry" {
+  content = jsonencode({
+    name               = var.participant_name,
+    url                = "http://${azurerm_container_group.edc.fqdn}:${local.edc_ids_port}",
+    supportedProtocols = ["ids-multipart"]
+  })
+  filename = "${path.module}/build/${var.participant_name}.json"
+}
+
+resource "azurerm_storage_share_file" "catalog_entry" {
```
Comment (marcgs) [inline deployment/terraform/main.tf]: With this solution there is no single public JSON file representing the registry anymore. I think we should consider having this, as it was taken as an assumption for further work as for example in https://github.com/agera-edc/MinimumViableDataspace/issues/61

Comment (marcgs) [reply deployment/terraform/main.tf]: Moving this comment down to the ADR doc

File: docs/developer/decision-records/2022-04-28-federated-catalog/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Federated Catalog
+
+## Decision
+
+Use files in an Azure file share to seed federated catalog of all participant EDC instances.
+
+The deployment pipeline for each participant creates a file in common folder in a file share, with a prefix corresponding to each unique deployment. For example, when deploying participants `company1` and `company2`, the files could be named:
```
Comment (marcgs) [inline docs/developer/decision-records/2022-04-28-federated-catalog/README.md:7]: With this solution there is no single public JSON file representing the registry anymore. I think we should consider having this, as it was taken as an assumption for further work as for example in https://github.com/agera-edc/MinimumViableDataspace/issues/61

File: docs/developer/decision-records/2022-04-28-federated-catalog/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+- `280-company1.json`
+- `280-company2.json`
```
Comment (pkirch) [inline docs/developer/decision-records/2022-04-28-federated-catalog/README.md:9]: Just to clarify: in this example the prefix `280` stands for a deployment?

File: .github/workflows/initialize.yaml [hunk 4]
```diff
@@ -38,3 +40,8 @@ jobs:
 
       - name: "Create Azure Container Registry"
         run: az acr create --resource-group "$COMMON_RESOURCE_GROUP" --name "$ACR_NAME" --sku Basic --admin-enabled -o table
+
+      - name: "Create Catalog Storage account"
```
Comment (marcgs) [inline .github/workflows/initialize.yaml]: Consider renaming it to "registry storage account" as it basically holds the registry information

File: docs/developer/decision-records/2022-04-28-federated-catalog/README.md (hunk 2 above)
Comment (marcgs) [reply docs/developer/decision-records/2022-04-28-federated-catalog/README.md:7]: As agreed, the requirement on the UI friendliness for the registry will be discussed in the coming refinement sessions. The current "catalog" storage account (consider renaming to registry) contains the same information spread over multiple json files. This comment can be resolved.

File: extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    void saveNodeEntries() {
+        var directory = mock(FederatedCacheNodeDirectory.class);
+        var sampleFile = getClass().getClassLoader().getResource("24-node1.json");
+        assertThat(sampleFile).isNotNull();
+        var nodeJsonDir = Path.of(sampleFile.getPath()).getParent();
+        var nodeJsonPrefix = "24-";
+        var monitor = new ConsoleMonitor();
+        var typeManager = new TypeManager();
+        var extension = new RefreshCatalogService(directory, nodeJsonDir, nodeJsonPrefix, monitor, typeManager);
```
Comment (marcgs) [inline extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java]: nit: rename to `service`?

File: extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        when(directory.getAll()).thenReturn(List.of(new FederatedCacheNode("node24-2", "", List.of(""))));
```
Comment (marcgs) [inline extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java]: Not sure to understand the effect of returning "node24-2" here. Shouldn't we test that already parsed nodes are not re-added?

File: system-tests/src/test/java/org/eclipse/dataspaceconnector/system/tests/utils/TransferSimulationUtils.java [hunk 7]
```diff
@@ -99,7 +98,7 @@ private static HttpRequestActionBuilder initiateContractNegotiation(String conne
                 .body(StringBody(loadContractAgreement(connectorAddress)))
                 .header(CONTENT_TYPE, "application/json")
                 .check(status().is(200))
-                .check(bodyString()
+                .check(jmesPath("id")
```
Comment (marcgs) [inline system-tests/src/test/java/org/eclipse/dataspaceconnector/system/tests/utils/TransferSimulationUtils.java:101]: how is this related to this change?

**Review by marcgs — APPROVED** [review]

File: extensions/refresh-catalog/src/main/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogService.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    void saveNodeEntries() {
+        try {
+            monitor.info("Refreshing catalog");
+            var files = Files.find(nodeJsonDir, 1,
+                    (path, attrs) -> path.toFile().getName().startsWith(nodeJsonPrefix));
+            List<FederatedCacheNode> existingNodes = nodeDirectory.getAll();
+            files
+                    .map(this::parseFederatedCacheNode)
+                    .filter(n -> existingNodes.stream().noneMatch(m -> Objects.equals(m.getName(), n.getName())))
```
Comment (Izzzu) [inline extensions/refresh-catalog/src/main/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogService.java]: So we don't insert a node if there's already existing a node with the same name. Shouldn't we update the node in this case instead of skipping it?

File: system-tests/src/test/java/org/eclipse/dataspaceconnector/system/tests/utils/TransferSimulationUtils.java (hunk 7 above)
Comment (algattik) [reply system-tests/src/test/java/org/eclipse/dataspaceconnector/system/tests/utils/TransferSimulationUtils.java:101]: This was changed in upstream EDC

File: extensions/refresh-catalog/src/main/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogService.java (hunk 8 above)
Comment (algattik) [reply extensions/refresh-catalog/src/main/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogService.java]: The directory currently doesn't support that. Good point for a new EDC issue though

**Review by Izzzu — APPROVED** [review]

File: docs/developer/decision-records/2022-04-28-federated-catalog/README.md (hunk 2 above)
Comment (algattik) [reply docs/developer/decision-records/2022-04-28-federated-catalog/README.md:7]: Created #83

**Comment by algattik** [comment]
"Initialize CD" workflow run at https://github.com/agera-edc/MinimumViableDataspace/actions/runs/2249407496

File: .github/workflows/initialize.yaml (hunk 4 above)
Comment (algattik) [reply .github/workflows/initialize.yaml]: Renamed

File: docs/developer/decision-records/2022-04-28-federated-catalog/README.md (hunk 3 above)
Comment (algattik) [reply docs/developer/decision-records/2022-04-28-federated-catalog/README.md:9]: Clarified in file

File: extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java (hunk 6 above)
Comment (algattik) [reply extensions/refresh-catalog/src/test/java/org/eclipse/dataspaceconnector/mvd/RefreshCatalogServiceTest.java]: Right, in fact in this version it runs only once and is the only actor loading nodes. So I've removed the duplicate detection logic from the test and implementation.
